### ISSUE TITLE
About gradient accumulation 

### ISSUE THREAD

**Issue by Rapisurazurite** [lukas-blecher/LaTeX-OCR#222]
Hello, Here is my train result, use default config file and batchsize 64 and microbatch -1
![image](https://user-images.githubusercontent.com/43494437/205339648-60458093-eb0a-4e2e-a2ee-9c202677f94d.png)
However, the BLUE score is far from your pretrained checkpoint.
I noticed that this project uses micro-batch hyperparameter to reduce gpu memory usage,
```
  for j in range(0, len(im), microbatch):
      tgt_seq, tgt_mask = seq['input_ids'][j:j+microbatch].to(device), seq['attention_mask'][j:j+microbatch].bool().to(device)
      loss = model.data_parallel(im[j:j+microbatch].to(device), device_ids=args.gpu_devices, tgt_seq=tgt_seq, mask=tgt_mask)*microbatch/args.batchsize
      loss.backward()  # data parallism loss is a vector
      total_loss += loss.item()
      torch.nn.utils.clip_grad_norm_(model.parameters(), 1)
  opt.step()
  scheduler.step()
```
It seems the loss should be rescaled when added to total_loss ( loss.item() / num_microbatch )?
Is this the reason why I reproduced worse? Because my equivalent learning rate is smaller?

Thanks.

**Comment by lukas-blecher** [comment]
Hi, sorry for the late reply.

I think the reason for the worse performance is that I trained the model for longer than specified in the config file. There was also a lot of babysitting while training, meaning I had to restart sometimes and played around with the learning rate.

On the micro batch issue, I considered normalizing the loss at some point but decided against it. Can't find the reference right now.
But I think you are right with the effective learning rate. What I implemented is more like gradient accumulation.
Either way it shouldn't make a difference since you set the microbatch to -1 and thus disabling the functionality.
The pretrained model was trained before I implemented it, so it should be equivalent.
